RecoPixelVertexing/PixelTriplets/src/CellularAutomaton.cc

```diff
@@ -1,265 +1,210 @@
+  float region_origin_y = region.origin().y();
+  float region_origin_radius = region.originRBound();
+
+  std::vector<bool> alreadyVisitedLayerPairs;
+  alreadyVisitedLayerPairs.resize(theLayerGraph.theLayerPairs.size());
+  for (auto visited : alreadyVisitedLayerPairs) {
+    visited = false;
+  }
+  for (int rootVertex : theLayerGraph.theRootLayers) {
+    std::queue<int> LayerPairsToVisit;
+
+    for (int LayerPair : theLayerGraph.theLayers[rootVertex].theOuterLayerPairs) {
+      LayerPairsToVisit.push(LayerPair);
+    }
 
+    unsigned int numberOfLayerPairsToVisitAtThisDepth = LayerPairsToVisit.size();
+
+    while (not LayerPairsToVisit.empty()) {
+      auto currentLayerPair = LayerPairsToVisit.front();
+      auto & currentLayerPairRef  = theLayerGraph.theLayerPairs[currentLayerPair];
+      auto & currentInnerLayerRef = theLayerGraph.theLayers[currentLayerPairRef.theLayers[0]];
+      auto & currentOuterLayerRef = theLayerGraph.theLayers[currentLayerPairRef.theLayers[1]];
+      bool allInnerLayerPairsAlreadyVisited{true};
+
+      for (auto innerLayerPair : currentInnerLayerRef.theInnerLayerPairs) {
+        allInnerLayerPairsAlreadyVisited &= alreadyVisitedLayerPairs[innerLayerPair];
+      }
+
+      if (alreadyVisitedLayerPairs[currentLayerPair] == false and allInnerLayerPairsAlreadyVisited) {
+        const HitDoublets *doubletLayerPairId = hitDoublets[currentLayerPair];
+        auto numberOfDoublets = doubletLayerPairId->size();
+        currentLayerPairRef.theFoundCells[0] = cellId;
+        currentLayerPairRef.theFoundCells[1] = cellId + numberOfDoublets;
+        for (unsigned int i = 0; i < numberOfDoublets; ++i) {
+          allCells.emplace_back(doubletLayerPairId, i,
+                                doubletLayerPairId->innerHitId(i),
+                                doubletLayerPairId->outerHitId(i));
+
+          currentOuterLayerRef.isOuterHitOfCell[doubletLayerPairId->outerHitId(i)].push_back(cellId);
+
+          cellId++;
+
+          auto & neigCells = currentInnerLayerRef.isOuterHitOfCell[doubletLayerPairId->innerHitId(i)];
+          allCells.back().checkAlignmentAndTag(
+              allCells, neigCells, ptmin, region_origin_x, region_origin_y,
+              region_origin_radius, thetaCut, phiCut, hardPtCut);
+        }
+        assert(cellId == currentLayerPairRef.theFoundCells[1]);
+        for (auto outerLayerPair : currentOuterLayerRef.theOuterLayerPairs) {
+          LayerPairsToVisit.push(outerLayerPair);
+        }
+
+        alreadyVisitedLayerPairs[currentLayerPair] = true;
+      }
+      LayerPairsToVisit.pop();
+      numberOfLayerPairsToVisitAtThisDepth--;
+      if (numberOfLayerPairsToVisitAtThisDepth == 0) {
+        numberOfLayerPairsToVisitAtThisDepth = LayerPairsToVisit.size();
+      }
+    }
+  }
 }
 
 void CellularAutomaton::evolve(const unsigned int minHitsPerNtuplet)
 {
   allStatus.resize(allCells.size());
-  
-  
+
   unsigned int numberOfIterations = minHitsPerNtuplet - 2;
```

Comment:
Ok, minHitsPerNtuplet is defined constant to "4" in the calling class in
http://cmslxr.fnal.gov/source/RecoPixelVertexing/PixelTriplets/src/CAHitQuadrupletGenerator.cc#0188

However, I find potentially dangerous to make differences of ```unsigned int```'s: is it fixed by the logic of the code that numberOfIterations cannot ever be passed as less or equal 2?

(Not strictly related to your PR, but I hit it while reviewing and I think it is worth to ask)

Reply:
I wouldn't know...
@felicepantaleo can you comment ?